### PR TITLE
Customizable Headers and Footers, fixes #11

### LINKED ISSUE #11: Headers/Footers
**Enhancement** 

In dgrid 1, the header is fully managed by the widget itself, while the footer is customizable - used by the pagination mixin and for user-provided content like summary rows (http://dgrid.io/tutorials/1.0/summary_row/).

Users have requested customization of the header in the same way (https://github.com/SitePen/dgrid/issues/211) as the footer. Making both areas behave similarly is something we should pursue in dgrid 2.

Both the header and the footer should operate both with fixed positioning as is the case with pagination or locked to the horizontal scrolling of the grid - or column sets that scroll independently - as is the case with summary rows and headers.

## Questions

* Should there be "more than 1" header or footer such as having both pagination and a summary row?
* Should these be configurable?
* Should we have bundled into the grid all header variations, pagination widgets, and summary widgets?
* If we do allow for configuration, how are these passed to the grid?
* What would keeping horizontal scrolling in sync look like?

### PR DESCRIPTION
**Pull request by pottedmeat**
**Type:** feature

The following has been addressed in the PR:

* [ ] There is a related issue
* [ ] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [ ] Unit or Functional tests are included in the PR

**Description:**

* Header => ColumnHeaders
* HeaderCell => ColumnHeaderCell
* Adds Header as a general-purpose wrapper, like footer
* Adds headers/footers as grid properties to allow custom children
* Provides an enum to indicate you want column headers in the header or footer area

### CODE REVIEW

File: src/ColumnHeaders.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import { v, w } from '@dojo/widget-core/d';
+import { DNode } from '@dojo/widget-core/interfaces';
+import { RegistryMixin, RegistryMixinProperties }  from '@dojo/widget-core/mixins/Registry';
+import { ThemeableMixin, theme, ThemeableProperties } from '@dojo/widget-core/mixins/Themeable';
```
Comment (msssk) [inline src/ColumnHeaders.ts]: `theme` should be first

**Review by msssk — APPROVED** [review]

File: src/Grid.ts [hunk 2]
```diff
@@ -25,6 +27,8 @@ export const GridBase = ThemeableMixin(RegistryMixin(WidgetBase));
 export interface GridProperties extends ThemeableProperties, HasColumns {
 	registry?: GridRegistry;
 	dataProvider: DataProviderBase;
+	footers?: Array<HeaderType | DNode>;
```
Comment (kitsonk) [inline src/Grid.ts]: generally we use `(HeaderType | DNode)[]`

File: src/Grid.ts [hunk 3]
```diff
@@ -79,18 +85,35 @@ class Grid extends GridBase<GridProperties> {
...
 		}, [
 			w<Header>('header', {
-				columns,
 				registry,
-				sortDetails,
-				theme,
-				onSortRequest
-			}),
+				theme
+			}, <DNode[]> headers.map((child) => {
```
Comment (kitsonk) [inline src/Grid.ts]: Seems odd we need a cast here.  Also this logic is quite hard to follow in line and is duplicated below.  This should probably be broken into a function that is easier to grok and maintain.

**Review by kitsonk — CHANGES_REQUESTED** [review]

**Review by kitsonk — APPROVED** [review]
